File: src/main/java/io/github/mg138/tsbook/items/data/IdentificationTag.java
package io.github.mg138.tsbook.items.data;

import com.google.gson.JsonSyntaxException;

import io.github.mg138.tsbook.Book;
import io.github.mg138.tsbook.items.ItemIdentification;

import org.bukkit.persistence.PersistentDataAdapterContext;
import org.bukkit.persistence.PersistentDataType;
import org.jetbrains.annotations.NotNull;

import java.io.*;

public class IdentificationTag implements PersistentDataType<String, ItemIdentification> {
    @NotNull
    @Override
    public Class<String> getPrimitiveType() {
        return String.class;
    }

    @NotNull
    @Override
    public Class<ItemIdentification> getComplexType() {
        return ItemIdentification.class;
    }

    @NotNull
    @Override
    public String toPrimitive(@NotNull ItemIdentification itemIdentification, @NotNull PersistentDataAdapterContext persistentDataAdapterContext) {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        String result = null;
        try {
            ObjectOutputStream oos = new ObjectOutputStream(bos);
            result = Book.Companion.getGson().toJson(itemIdentification);
            oos.writeObject(result);
            oos.flush();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                bos.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        assert result != null;
        return result;
    }

    @NotNull
    @Override
    public ItemIdentification fromPrimitive(@NotNull String string, @NotNull PersistentDataAdapterContext persistentDataAdapterContext) {
        ItemIdentification result = null;
        try {
            result = Book.Companion.getGson().fromJson(string, ItemIdentification.class);
        } catch (JsonSyntaxException e) {
            e.printStackTrace();
        }
        assert result != null;
        return result;
    }
}

File: src/main/kotlin/io/github/mg138/tsbook/item/storage/IdentificationTag.kt
package io.github.mg138.tsbook.item.storage

import com.google.gson.JsonSyntaxException
import io.github.mg138.tsbook.Book.Companion.gson
import io.github.mg138.tsbook.item.data.Identification
import org.bukkit.persistence.PersistentDataAdapterContext
import org.bukkit.persistence.PersistentDataType
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectOutputStream

object IdentificationTag : PersistentDataType<String, Identification> {
    override fun getPrimitiveType(): Class<String> {
        return String::class.java
    }

    override fun getComplexType(): Class<Identification> {
        return Identification::class.java
    }

    override fun toPrimitive(
        identification: Identification,
        persistentDataAdapterContext: PersistentDataAdapterContext
    ): String {
        val bos = ByteArrayOutputStream()
        var result: String? = null
        try {
            val oos = ObjectOutputStream(bos)
            result = gson.toJson(identification)
            oos.writeObject(result)
            oos.flush()
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            try {
                bos.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        assert(result != null)
        return result!!
    }

    override fun fromPrimitive(
        string: String,
        persistentDataAdapterContext: PersistentDataAdapterContext
    ): Identification {
        var result: Identification? = null
        try {
            result = gson.fromJson(string, Identification::class.java)
        } catch (e: JsonSyntaxException) {
            e.printStackTrace()
            println("The error string: $string")
        }
        assert(result != null)
        return result!!
    }
}